### ISSUE TITLE
Incorrect gem installed for platform

### ISSUE THREAD

**Issue by bjfish** [rubygems/rubygems#5508]
### Describe the problem as clearly as you can

During `bundle install`, the `musl` version of a gem is installed instead of the expected platform version.

Output:
```
Using bundler 2.3.12
Fetching libv8-node 16.10.0.0 (x86_64-linux-musl)
Installing libv8-node 16.10.0.0 (x86_64-linux-musl)
Bundle complete! 1 Gemfile dependency, 2 gems now installed.

$ ruby -e 'p Gem.platforms'
["ruby", #<Gem::Platform:0x0000564d64ec3080 @cpu="x86_64", @os="linux", @version=nil>]

```

### Post steps to reproduce the problem

Gemfile
```
# frozen_string_literal: true

source "https://rubygems.org"

gem "libv8-node"

```

With the above `Gemfile`, 
1. run `bundle install`
3. `bundle lock --add-platform ruby`
4. `bundle lock --remove-platform x86_64-linux`
5. `gem uninstall libv8-node`
6. `bundle install`

The incorrect musl version above should be shown in the output.

### Which command did you run?

  bundle install

### What were you expecting to happen?

The musl version should not be installed. The gem version in the lockfile should be installed.

### What actually happened?

The musl version was installed.

## Environment

```
Bundler       2.3.12
  Platforms   ruby, x86_64-linux
Ruby          3.0.3p157 (2021-11-24 revision 3fb7d2cadc18472ec107b14234933b017a33c14d) [x86_64-linux]
  Full Path   /home/brandon/.rbenv/versions/3.0.3/bin/ruby
  Config Dir  /home/brandon/.rbenv/versions/3.0.3/etc
RubyGems      3.3.12
  Gem Home    /home/brandon/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0
  Gem Path    /home/brandon/.local/share/gem/ruby/3.0.0:/home/brandon/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0
  User Home   /home/brandon
  User Path   /home/brandon/.local/share/gem/ruby/3.0.0
  Bin Dir     /home/brandon/.rbenv/versions/3.0.3/bin
Tools         
  Git         2.30.2
  RVM         not installed
  rbenv       rbenv 1.1.2
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2022-04-20
Git SHA           75031e03f4
Released Version  true
```

### Gemfile

```ruby
# frozen_string_literal: true

source "https://rubygems.org"

gem "libv8-node"
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    libv8-node (16.10.0.0)

PLATFORMS
  ruby

DEPENDENCIES
  libv8-node

BUNDLED WITH
   2.3.12
```

**Comment by bjfish** [comment]
This may be related to: https://github.com/rubyjs/libv8-node/issues/5

**Comment by bjfish** [comment]
Feel free to close this as duplicate. I assume it is a duplicate of the above mentioned issue which links to this PR: https://github.com/rubygems/rubygems/pull/4488

**Comment by eregon** [comment]
Also related, this issue: https://github.com/rubygems/rubygems/issues/3174 which had a PR which got reverted (https://github.com/rubygems/rubygems/pull/4082).
@deivid-rodriguez How can we solve this?
Note this issue applies to CRuby and TruffleRuby, it's not TruffleRuby-specific.

**Comment by deivid-rodriguez** [comment]
Can someone verify that #4488 fixes this problem? My Linux laptop decided to not boot anymore today, so I can't try this myself easily.

**Comment by eregon** [comment]
I can repro with bundler 2.3.19 and the instructions above.
And indeed with that PR it's fixed:
```
$ bundle _2.4.0.dev_ install
Fetching gem metadata from https://rubygems.org/.
Using bundler 2.4.0.dev
Fetching libv8-node 16.10.0.0 (x86_64-linux)
Installing libv8-node 16.10.0.0 (x86_64-linux)
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
```

**Comment by deivid-rodriguez** [comment]
Closed by #5852!

**Comment by eregon** [comment]
This doesn't seem quite solved yet, or it regressed somehow: https://github.com/rubygems/rubygems/issues/6165
